PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/toolkit/intellij/appservice/component/input/ComboBoxAzureArtifact.java

```diff
@@ -0,0 +1,147 @@
+                               this.setItems(items);
+                               this.setLoading(false);
+                               this.resetDefaultValue(defaultArtifactType, defaultPath);
+                           }), (e) -> {
+                               this.handleLoadingError(e);
+                           });
+    }
+
+    @NotNull
+    @Override
+    protected List<? extends AzureArtifact> loadItems() throws Exception {
+        return AzureArtifactManager.getInstance(project).getAllSupportedAzureArtifacts();
+    }
+
+    @Nullable
+    @Override
+    protected ExtendableTextComponent.Extension getExtension() {
+        return ExtendableTextComponent.Extension.create(
+                AllIcons.General.OpenDisk, "Open file", this::onSelectFile);
+    }
+
+    protected String getItemText(Object item) {
+        if (item instanceof AzureArtifact) {
+            return String.format("%s : %s", ((AzureArtifact) item).getType(), ((AzureArtifact) item).getName());
+        } else {
+            return StringUtils.EMPTY;
+        }
+    }
+
+    @Nullable
+    protected Icon getItemIcon(Object item) {
+        return item instanceof AzureArtifact ? ((AzureArtifact) item).getIcon() : null;
+    }
+
+    private void onSelectFile() {
+        final FileChooserDescriptor fileDescriptor = FileChooserDescriptorFactory.createSingleFileDescriptor();
+        if (filter != null) {
+            fileDescriptor.withFileFilter(filter);
+        }
+        fileDescriptor.withTitle("Select artifact to deploy");
+        final VirtualFile file = FileChooser.chooseFile(fileDescriptor, null, null);
+        if (file != null && file.exists()) {
+            addOrSelectExistingVirtualFile(file);
+        }
+    }
+
+    private void addOrSelectExistingVirtualFile(VirtualFile virtualFile) {
+        final AzureArtifact selectArtifact = AzureArtifact.createFromFile(virtualFile);
+        final List<AzureArtifact> artifacts = UIUtils.listComboBoxItems(ComboBoxAzureArtifact.this);
+        final AzureArtifact existingArtifact =
+                artifacts.stream().filter(artifact -> artifact.getType() == AzureArtifactType.Artifact
+                        && StringUtils.equalsAnyIgnoreCase(artifact.getTargetPath(), selectArtifact.getTargetPath()))
+                         .findFirst().orElse(null);
+        if (existingArtifact == null) {
+            ComboBoxAzureArtifact.this.addItem(selectArtifact);
+            ComboBoxAzureArtifact.this.setSelectedItem(selectArtifact);
+        } else {
+            ComboBoxAzureArtifact.this.setSelectedItem(existingArtifact);
+        }
+    }
+
+    private void resetDefaultValue(final AzureArtifactType defaultArtifactType, final String defaultPath) {
+        final List<AzureArtifact> artifacts = UIUtils.listComboBoxItems(this);
+        final AzureArtifact defaultArtifact =
+                artifacts.stream().filter(artifact -> StringUtils.equals(artifact.getTargetPath(), defaultPath)
```

Comment:
AzureArtifact is designed to use getArtifactIdentifier to detect same object,  see PR https://github.com/microsoft/azure-tools-for-java/pull/4655/files